doc/mgr/balancer.rst

```diff
@@ -0,0 +1,285 @@
+balancer plugin
+================
+
+Typically, if you say, the system is 70% filled, then, all the OSDs,
+should be 70% filled (roughly). It shouldn’t be the case that, a particular
+OSD is 90% filled whereas another one is 20% filled. If such a case exist,
+then we say there is an uneven distribution. Uneven distribution may arise
+out of a number of reasons, including improper distribution of PGs among the
+OSDs. Even if the PGs, were balanced properly, with respect to the weights
+of the OSDs, there is still chance of uneven distribution, because the number
+of objects assigned to different PGs are not in required proportion. And even
+if they were, the sizes of each object might vary.
+
+The balancer plugin will automatically optimize the pg (or objects or byte-level)
+distribution.  One needs to just turn it on and it will slowly and continuously
+optimize the layout without the user having to think about it. In this case, optimization
+means, changing the weights of the devices (OSDs), so that the right amount of PGs, get
+mapped to each device, with respect to their relative weights. If one wanted to attain
+byte-level even-distribution, they would have to change the key (``ceph balancer key``)
+to `byte`. It is however, suggested that the key be set to `auto` mode, with weights for
+each parameter.
+
+While it is recommended that the balancer module be best left to run auto-magically,
+it might interest certain users to turn off the automatic optimization, and instead
+create optimization plans themselves, and execute them, whenever they feel the need
+to do so. While there is no fixed way to determine when to execute the rebalance
+plan, the *score* of the system on each parameter and the average score (can be
+viewed using ``balancer eval``) will give a general idea as to how uneven the
+current distribution is. The lower the score, the better the distribution,
+with 0 meaning perfect distribution, and 1 meaning the average over-utilized device
+is infinitely more weighted than the average utilization of all devices.
+
+As a thumb-rule, if score is 0.08, then the average over-utilized device is 10%
+over-weighted compared to the average utilization of all OSDs, and a score of 0.68
+implies 100% more weighted. The scoring is non-linear, to signify that presence of
+highly over-utilizaed devices is a pretty bad situation.
+
+Wokring Process and Performance
+-------------------------------
+
+At each instance of optimization, balancer module creates a `Plan`. Then it uses the
+CRUSH to identify the various OSDs in a pool and calculate the distribution of `pg`,
+`bytes` and `objects`, from the OSDMap. Then it calculates, how the target distribution
+of each of the parameter(`pg`, `bytes` and `objects`). Using an iterative gradient
```

Comment:
is

> Using an iterative gradient  descent algorithm

a complete sentence or a part of it?

Reply:
It was supposed to be a complete sentence, but I guess, it became a bit informal. Updated.